metrics/internal/lv/space.go

```diff
@@ -77,25 +86,39 @@ type node struct {
 type pair struct{ label, value string }
 
 func (n *node) observe(lvs LabelValues, value float64) {
-	n.mtx.Lock()
-	defer n.mtx.Unlock()
 	if len(lvs) <= 0 {
+		n.mtx.Lock()
 		n.observations = append(n.observations, value)
+		n.mtx.Unlock()
 		return
 	}
 	if len(lvs) < 2 {
 		panic("too few LabelValues; programmer error!")
 	}
 	head, tail := pair{lvs[0], lvs[1]}, lvs[2:]
+	n.getChild(head).observe(tail, value)
+}
+
+func (n *node) getChild(head pair) *node {
+	// First try without taking write lock.
+	n.mtx.RLock()
+	child, ok := n.children[head]
+	n.mtx.RUnlock()
+	if ok {
+		return child
+	}
+	// No luck, we need to go through the slow path.
+	n.mtx.Lock()
+	defer n.mtx.Unlock()
 	if n.children == nil {
 		n.children = map[pair]*node{}
 	}
-	child, ok := n.children[head]
+	child, ok = n.children[head]
```

Comment:
why is this check here, we already know that this will be `!ok`

Reply:
We don't know that. It can happen that between the time we do `n.mtx.RUnlock()` and `n.mtx.Lock()` someone else has created the node. I've added a comment.

Reply:
uff.. well i missed that - that's why i felt like having all these methods atomic would be more fool-proof